core/llm_google.go

```diff
@@ -164,91 +122,193 @@ func (c *GenaiClient) SendQuery(ctx context.Context, history []ModelMessage, too
+func (c *GenaiClient) processStreamResponse(
+	ctx context.Context,
+	stream *genai.GenerateContentResponseIterator,
+	onTextPart textPartHandler,
+	onTokenUsage tokenUsageHandler,
+) (content string, toolCalls []ToolCall, tokenUsage LLMTokenUsage, err error) {
+
+	content = ""
+	toolCalls = []ToolCall{}
+	tokenUsage = LLMTokenUsage{}
 
-	var content string
-	var toolCalls []ToolCall
-	var tokenUsage LLMTokenUsage
 	for {
-		res, err := stream.Next()
-		if err != nil {
-			if errors.Is(err, iterator.Done) {
+		res, streamErr := stream.Next()
+		if streamErr != nil {
+			if errors.Is(streamErr, iterator.Done) {
 				break
 			}
-			if apiErr, ok := err.(*apierror.APIError); ok {
-				// unwrap the APIError
-				return nil, fmt.Errorf("google API error occurred: %w", apiErr.Unwrap())
+			if apiErr, ok := streamErr.(*apierror.APIError); ok {
+				err = fmt.Errorf("google API error occurred: %w", apiErr.Unwrap())
+				return content, toolCalls, tokenUsage, err
 			}
-			return nil, err
+
+			err = streamErr
+			return content, toolCalls, tokenUsage, err
 		}
 
-		// Keep track of the token usage
 		if res.UsageMetadata != nil {
-			if res.UsageMetadata.CandidatesTokenCount > 0 {
-				count := int64(res.UsageMetadata.CandidatesTokenCount)
-				tokenUsage.OutputTokens += count
-				tokenUsage.TotalTokens += count
-				outputTokens.Record(ctx, count, metric.WithAttributes(attrs...))
-			}
-			if res.UsageMetadata.PromptTokenCount > 0 {
-				count := int64(res.UsageMetadata.PromptTokenCount)
-				tokenUsage.InputTokens += count
-				tokenUsage.TotalTokens += count
-				inputTokens.Record(ctx, count, metric.WithAttributes(attrs...))
-			}
-			if res.UsageMetadata.CachedContentTokenCount > 0 {
-				count := int64(res.UsageMetadata.CachedContentTokenCount)
-				inputTokensCacheReads.Record(ctx, count, metric.WithAttributes(attrs...))
+			if err := onTokenUsage(ctx, res.UsageMetadata, &tokenUsage); err != nil {
+				err = fmt.Errorf("token usage handler failed: %w", err)
+				return content, toolCalls, tokenUsage, err
 			}
 		}
 
 		if len(res.Candidates) == 0 {
-			return nil, &ModelFinishedError{
-				Reason: "no response from model",
-			}
+			err = &ModelFinishedError{Reason: "no response from model"}
+			return content, toolCalls, tokenUsage, err
 		}
 		candidate := res.Candidates[0]
 		if candidate.Content == nil {
-			return nil, &ModelFinishedError{
-				Reason: candidate.FinishReason.String(),
-			}
+			err = &ModelFinishedError{Reason: candidate.FinishReason.String()}
+			return content, toolCalls, tokenUsage, err
 		}
 
 		for _, part := range candidate.Content.Parts {
 			switch x := part.(type) {
 			case genai.Text:
-				fmt.Fprint(stdio.Stdout, x)
-				content += string(x)
+				textChunk := string(x)
+				if err := onTextPart(ctx, textChunk); err != nil {
+					err = fmt.Errorf("text part handler failed: %w", err)
+					return content, toolCalls, tokenUsage, err
+				}
+				content += textChunk
```

Comment:
No need for onTextPart, just pass an `stdout io.Writer` to `processStreamResponse`.
```suggestion
				fmt.Fprint(stdout, x)
				content += string(x)
```